carstore/read_only_blockstore.go

```diff
@@ -0,0 +1,83 @@
+package carstore
+
+import (
+	"io"
+	"sync"
+
+	bstore "github.com/ipfs/go-ipfs-blockstore"
+	carv2 "github.com/ipld/go-car/v2"
+	"github.com/ipld/go-car/v2/blockstore"
+	"golang.org/x/xerrors"
+)
+
+type ClosableBlockstore interface {
+	bstore.Blockstore
+	io.Closer
+}
+
+// CarReadOnlyStoreTracker tracks the lifecycle of a ReadOnly CAR Blockstore and makes it easy to create/get/cleanup the blockstores.
+// It's important to close a CAR Blockstore when done using it so that the backing CAR file can be closed.
+type CarReadOnlyStoreTracker struct {
+	mu     sync.RWMutex
+	stores map[string]ClosableBlockstore
+}
+
+func NewReadOnlyStoreTracker() *CarReadOnlyStoreTracker {
+	return &CarReadOnlyStoreTracker{
+		stores: make(map[string]ClosableBlockstore),
+	}
+}
+
+func (r *CarReadOnlyStoreTracker) Add(key string, bs ClosableBlockstore) (bool, error) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	if _, ok := r.stores[key]; ok {
+		return false, nil
+	}
+
+	r.stores[key] = bs
+	return true, nil
+}
+
+func (r *CarReadOnlyStoreTracker) GetOrCreate(key string, carFilePath string) (ClosableBlockstore, error) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	if bs, ok := r.stores[key]; ok {
+		return bs, nil
+	}
+
+	rdOnly, err := blockstore.OpenReadOnly(carFilePath, carv2.ZeroLengthSectionAsEOF(true), blockstore.UseWholeCIDs(true))
+	if err != nil {
+		return nil, xerrors.Errorf("failed to open read-only blockstore: %w", err)
+	}
+	r.stores[key] = rdOnly
+
+	return rdOnly, nil
+}
+
+func (r *CarReadOnlyStoreTracker) Get(key string) (ClosableBlockstore, error) {
+	r.mu.RLock()
+	defer r.mu.RUnlock()
+
+	if bs, ok := r.stores[key]; ok {
+		return bs, nil
+	}
+
+	return nil, xerrors.Errorf("could not get blockstore for key %s: %w", key, ErrNotFound)
+}
+
+func (r *CarReadOnlyStoreTracker) CleanBlockstore(key string) error {
```

Comment:
This is a bit of a misnomer because it's not "cleaning" anything; it's just untracking it.

Reply:
Will rename.